rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -520,7 +747,10 @@ pub enum ExecutionTask {
 
     /// On low Wasm memory hook.
     /// The task exists only within an execution round, it never gets serialized.
-    OnLowWasmMemory,
+    /// If the encapsulated field is `None` that represents that hook condition was
+    /// not checked. Otherwise, the value inside `Some` represents the outcome of
+    /// the checking for hook condition.
```

Comment:
Wouldn't it be simpler to just not call `TaskQueue::push_front()` in that case? This is a `push` API that says "if the argument is `Some`, enqueue something; if it's `None`, do absolutely nothing".

Reply:
If I understand your comment correctly, that was exactly the behavior of the `TaskQueue` we did not call `push_front` for this case, but instead `set_status` was used. And then after your request to use more like `queue` style interface, that was changed.

Reply:
I was thinking about that, but after thinking about what `pop_front` will return in the case of `OnLowWasmMemory`, I decided to keep it as the `Optiona<bool>` so we can when calling `pop_front` return `OnLowWasmMemory(None)`. Because that is more intuitive than `OnLowWasmMemory(true)` of `OnLowWasmMemory(false)` since we did not perform any new check of the hook condition when poping from the queue.

Reply:
I suppose I would have to see the follow-up PR (which I won't as I'll be on vacation for the next two weeks) to understand this.

But what's the problem with `pop_front()` returning `OnLowWasmMemory(_)` when the task needs to be executed and returning `None` (or the next task) when the task should not be executed? What would the scheduler (or whoever is supposed to execute this task) be expected to do with `Some(true)` vs `Some(false)` vs `None`?

Reply:
That field only tells if the hook status should be updated when pushing on the queue. So no other part of the code should use it except the TaskQueue, and no other part should be expected to do anything with it, so it makes sense to return None always when poping from queue. Every time `OnLowWasmMemory(_)` is poped from the queue it means it should be executed.

The part of code that you are asking for will look something like:

```
struct TaskQueue{
...
fn pop_front(&mut self) -> Option<ExecutionTask> {
    ...
    if(slef.on_low_wasm_memory_hook_status = OnLowWasmMemoryHookStatus::Ready){
       slef.on_low_wasm_memory_hook_status = OnLowWasmMemoryHookStatus::Executed;
       return Some(ExecutionTask:: OnLowWasmMemory(None));
    }
   ...
}
...
}
```

Reply:
That's quite a convoluted API, though.

You could, of course, argue that `OnLowWasmMemory(None)` makes more sense than `OnLowWasmMemory(Some(true))` as an output of `pop_front()`. You could also argue the opposite, since the value inside the variant is meaningless to someone consuming it.

If you look at heartbeats and timers, the API is simply `enqueue(Heartbeat)`; `pop_front()` may return `Heartbeat`; and then `remove_heartbeat_and_global_timer()`. Why not do something similar here? Calling `enqueue(OnLowWasmMemory(Some(false))` with the semantic of "remove OnLowWasmMemory task" is already a warning sign by itself.

So you could have whoever perform the check either call `enqueue(OnLowWasmMemory)` or `remove(OnLowWasmMemory)` / `remove_on_low_wasm_memory()`, depending on the outcome of the check. And do nothing if they didn't do a check. Then, on the consumer end, you just `pop_front()` an `OnLowWasmMemory` (with no wrapped value). Or nothing, if the status is either `ConditionNotSatisfied` or `Executed`.

Reply:
> I suppose I would have to see the follow-up PR (which I won't as I'll be on vacation for the next two weeks) to understand this.

Given you are going to be on vacation for the next two weeks (and obviously we cannot be waiting for you to come back so we can merge), is

> So you could have whoever perform the check either call enqueue(OnLowWasmMemory) or remove(OnLowWasmMemory) / remove_on_low_wasm_memory(), depending on the outcome of the check. And do nothing if they didn't do a check. Then, on the consumer end, you just pop_front() an OnLowWasmMemory (with no wrapped value). Or nothing, if the status is either ConditionNotSatisfied or Executed.

a strong opinion or a nice to have for a better API?

Reply:
Ok, I modified it to use remove/enqueue logic. @alin-at-dfinity can you PTAL and approve?